HPX_SERIALIZATION_SPLIT_FREE does not stop compiler from looking for serialize() method

I want to enable HPX serialization of a library type, whose definition I cannot change. For this purpose I define a pair of free functions `save()`/`load()` and make use of `HPX_SERIALIZATION_SPLIT_FREE`. Unfortunately, the type in question has a method called `serialize()`, which has nothing to do with HPX and it apparently gets into the way of name lookup.

```c++
#include <hpx/runtime/serialization/serialize.hpp>

struct has_serialize {
  int i;
  double d;

  void serialize() {
    // Does something unrelated to HPX serialization ...
  }
};

template <typename Archive>
void save(Archive &ar, has_serialize const &hs, const unsigned int) {
  std::cout << "saving" << std::endl;
}

template <typename Archive>
void load(Archive &ar, has_serialize &hs, const unsigned int) {
  std::cout << "loading" << std::endl;
}

HPX_SERIALIZATION_SPLIT_FREE(has_serialize);

int main() {

  has_serialize in, out;

  std::vector<char> buf;

  // Serialize
  hpx::serialization::output_archive oa(buf, hpx::serialization::disable_data_chunking);
  oa << in;

  // Deserialize
  hpx::serialization::input_archive ia(buf);
  ia >> out;

  return 0;
}
```
```
include/hpx/runtime/serialization/access.hpp:88:21: error: no matching function for call to ‘has_serialize::serialize(hpx::serialization::input_archive&, int)’
                     const_cast<typename util::decay<T>::type&>(
                     ^~~~~~~~~~
hpx_bug.cpp:7:8: note: candidate: void has_serialize::serialize()
   void serialize() {
        ^~~~~~~~~
hpx_bug.cpp:7:8: note:   candidate expects 0 arguments, 2 provided
```
This example compiles if I comment out `has_serialize::serialize()`.

`HPX_SERIALIZATION_SPLIT_FREE_TEMPLATE` seems to be affected too.

It's worth noting that Boost.Serialization does not suffer from this issue.

--- a/hpx/runtime/serialization/access.hpp
+++ b/hpx/runtime/serialization/access.hpp
@@ -8,19 +8,17 @@
 #define HPX_SERIALIZATION_ACCESS_HPP
 
 #include <hpx/runtime/serialization/serialization_fwd.hpp>
-#include <hpx/traits/has_member_xxx.hpp>
 #include <hpx/traits/polymorphic_traits.hpp>
 #include <hpx/util/decay.hpp>
 
 #include <string>
 #include <type_traits>
+#include <utility>
 
 namespace hpx { namespace serialization
 {
     namespace detail
     {
-        HPX_HAS_MEMBER_XXX_TRAIT_DEF(serialize);
-
         template <class T> HPX_FORCEINLINE
         void serialize_force_adl(output_archive& ar, const T& t, unsigned)
         {
@@ -36,6 +34,26 @@ namespace hpx { namespace serialization
 
     class access
     {
+        template <class T>
+        class has_serialize
+        {
+            template <class T1> static std::false_type test(...);
+
+            // the following expression sfinae trick
+            // appears to work on clang-3.4, gcc-4.9,
+            // icc-16, msvc-2017 (at least)
+            // note that this detection would have been much easier
+            // to implement if there hadn't been an issue with gcc:
+            // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=82478
+            template <class T1, class = decltype(
+                    std::declval<typename std::remove_const<T1>::type &>().
+                        serialize(std::declval<output_archive &>(), 0u))>
+            static std::true_type test(int);
+
+        public:
+            static constexpr bool value = decltype(test<T>(0))::value;
+        };
+
         template <class T>
         class serialize_dispatcher
         {
@@ -73,7 +91,7 @@ namespace hpx { namespace serialization
             struct empty
             {
                 template <class Archive>
-                static void call(Archive& ar, T& t, unsigned)
+                static void call(Archive&, T&, unsigned)
                 {
                 }
             };
@@ -95,7 +113,7 @@ namespace hpx { namespace serialization
                 hpx::traits::is_intrusive_polymorphic<T>::value,
                 intrusive_polymorphic,
                 typename std::conditional<
-                    detail::has_serialize<T>::value,
+                    has_serialize<T>::value,
                     intrusive_usual,
                     typename std::conditional<
                         std::is_empty<T>::value,

--- a/hpx/util/tuple.hpp
+++ b/hpx/util/tuple.hpp
@@ -306,7 +306,7 @@ namespace hpx { namespace util
             }
 
             template <typename Archive>
-            void serialize(Archive& ar, unsigned int const version)
+            void serialize(Archive& ar, unsigned int const)
             {
                 int const _sequencer[] = {
                     ((ar & this->get<Is>()), 0)...

--- a/tests/unit/serialization/CMakeLists.txt
+++ b/tests/unit/serialization/CMakeLists.txt
@@ -21,6 +21,7 @@ set(tests
     serialization_partitioned_vector
     serialization_variant
     serialize_buffer
+    serialize_with_incompatible_signature
     zero_copy_serialization
 )
 

--- a/tests/unit/serialization/serialize_with_incompatible_signature.cpp
+++ b/tests/unit/serialization/serialize_with_incompatible_signature.cpp
@@ -1,0 +1,45 @@
+//  Copyright (c) 2018 Anton Bikineev
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/runtime/serialization/serialize.hpp>
+#include <hpx/util/lightweight_test.hpp>
+
+#include <vector>
+
+struct A
+{
+    double a;
+    int p;
+
+    void serialize(int, unsigned) {
+        // 3rd-party logic...
+    }
+};
+
+template <class Ar>
+void serialize(Ar& ar, A& a, unsigned) {
+    ar & a.a;
+    ar & a.p;
+}
+
+int main()
+{
+    std::vector<char> vector;
+    {
+        hpx::serialization::output_archive oar(vector);
+        A a{2., 4};
+        oar << a;
+    }
+
+    {
+        A a;
+        hpx::serialization::input_archive iar(vector);
+        iar >> a;
+        HPX_TEST_EQ(a.a, 2.);
+        HPX_TEST_EQ(a.p, 4);
+    }
+
+    return 0;
+}

